### PR TITLE
CARM implementation

### PR DESCRIPTION
**Pull request by A-Hilaly**
Issue #102 

Description of changes:

- [x] CARM implementation
- [x] Smoke tests
- [ ] Github action step

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/e2e/cross-account/smoke.sh [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+SCRIPTS_DIR="$ROOT_DIR/scripts"
+
+source "$SCRIPTS_DIR/lib/common.sh"
+source "$SCRIPTS_DIR/lib/k8s.sh"
+source "$SCRIPTS_DIR/lib/testutil.sh"
+
+
+# AWS_ACCOUNT_ID_2 should be configured to allow cross account resources management
+# from AWS_ACCOUNT_ID
+AWS_ACCOUNT_ID_2=""
```
Comment (A-Hilaly) [inline test/e2e/cross-account/smoke.sh:14]: I'm still working on this script. I need to figure out a way to switch accounts for github action steps

File: apis/core/v1alpha1/common.go [hunk 2]
```diff
@@ -21,3 +21,6 @@ type AWSAccountID string
 
 // AWSResourceName represents an AWS Resource Name (ARN)
 type AWSResourceName string
+
+// AWSRoleARN represents an AWS Role ARN
+type AWSRoleARN string
```
Comment (jaypipes) [inline apis/core/v1alpha1/common.go:26]: Just use the AWSResourceName as the data type, please.

File: pkg/runtime/session.go [hunk 3]
```diff
@@ -16,11 +16,19 @@ package runtime
...
 )
 
-func NewSession(region ackv1alpha1.AWSRegion) (*session.Session, error) {
+// NewSession returns a new session object. Buy default the returned session is
+// created using pod IRSA environment variables. If assumeRoleARN is not empty,
+// NewSession will call STS::AssumeRole and use the returned credentials to create
+// the session.
+func NewSession(
+	region ackv1alpha1.AWSRegion,
+	assumeRoleARN ackv1alpha1.AWSRoleARN,
```
Comment (jaypipes) [inline pkg/runtime/session.go:30]: just use `AWSResouceName` data type.

File: pkg/types/aws_resource_manager.go [hunk 4]
```diff
@@ -45,7 +48,12 @@ type AWSResourceManager interface {
 	// higher-level reonciler determines whether or not the desired differs
 	// from the latest observed and decides whether to call the resource
 	// manager's Update method
-	Update(context.Context, /*desired*/ AWSResource, /*latest*/ AWSResource, *ackcompare.Reporter) (AWSResource, error)
+	Update(
+		context.Context,
+		AWSResource, // desired
+		AWSResource, // latest
+		*ackcompare.Reporter,
+	) (AWSResource, error)
```
Comment (jaypipes) [inline pkg/types/aws_resource_manager.go:56]: FYI, it's better to leave style-only changes like this to a separate style-only PR. Makes reviewing the code easier since it separates functional code changes from non-functional code changes.

File: services/s3/config/crd/bases/s3.services.k8s.aws_buckets.yaml [hunk 5]
```diff
@@ -1,10 +1,10 @@
 
 ---
-apiVersion: apiextensions.k8s.io/v1beta1
+apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.3.1-0.20200716001835-4a903ddb7005
+    controller-gen.kubebuilder.io/version: (devel)
```
Comment (jaypipes) [inline services/s3/config/crd/bases/s3.services.k8s.aws_buckets.yaml:7]: hmm.... not sure this is correct.

File: services/s3/config/crd/bases/s3.services.k8s.aws_buckets.yaml [hunk 6]
```diff
@@ -15,122 +15,122 @@ spec:
...
+    subresources:
+      status: {}
 status:
   acceptedNames:
     kind: ""
     plural: ""
-  conditions: null
-  storedVersions: null
+  conditions: []
+  storedVersions: []
```
Comment (jaypipes) [inline services/s3/config/crd/bases/s3.services.k8s.aws_buckets.yaml:136]: all these changes to CRD YAML files should be in a separate PR. If these changes were from an updated kubebuilder, then we should have a separate PR that only updates the service controller generated files due to that upgraded KB version.

File: test/e2e/cross-account/smoke.sh [hunk 7]
```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env bash
```
Comment (jaypipes) [inline test/e2e/cross-account/smoke.sh:1]: Place this file in the `test/e2e/s3` directory and call it `cross-account.sh` :) That way, when we run the S3 e2e tests, we'll run the CARM tests.

File: test/e2e/cross-account/smoke.sh [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+source "$SCRIPTS_DIR/lib/common.sh"
+source "$SCRIPTS_DIR/lib/k8s.sh"
+source "$SCRIPTS_DIR/lib/testutil.sh"
+
+
+# AWS_ACCOUNT_ID_2 should be configured to allow cross account resources management
+# from AWS_ACCOUNT_ID
+AWS_ACCOUNT_ID_2=""
+DEFAULT_REGION="eu-west-2"
```
Comment (jaypipes) [inline test/e2e/cross-account/smoke.sh:15]: Note that there is already an `AWS_REGION` environment variable defined in the `kind-build-test.sh` script:

https://github.com/aws/aws-controllers-k8s/blob/b39842d4ce92595478c4e9e6af4a9667b33985e4/scripts/kind-build-test.sh#L20

So, probably best to call this something like `AWS_REGION_ALT`?

File: test/e2e/cross-account/smoke.sh (hunk 1 above)
Comment (jaypipes) [inline test/e2e/cross-account/smoke.sh:14]: How about calling this env var `AWS_ACCOUNT_ID_ALT`

Comment (jaypipes) [reply test/e2e/cross-account/smoke.sh:14]: Note that the `aws` CLI tool supports named profiles for exactly this purpose:

https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-profiles.html

Comment (A-Hilaly) [reply test/e2e/cross-account/smoke.sh:14]: Oh yeah let's use this!

File: pkg/types/aws_resource_manager.go (hunk 4 above)
Comment (A-Hilaly) [reply pkg/types/aws_resource_manager.go]: Agreed 👍 gofmt was automatically doing weird comment style change... i'll make this change in a different PR

File: services/s3/config/crd/bases/s3.services.k8s.aws_buckets.yaml (hunk 5 above)
Comment (A-Hilaly) [reply services/s3/config/crd/bases/s3.services.k8s.aws_buckets.yaml]: My bad, i was used the latest controller-gen version, i just regenerated with `v3.1.0`

File: services/s3/config/crd/bases/s3.services.k8s.aws_buckets.yaml (hunk 6 above)
Comment (A-Hilaly) [reply services/s3/config/crd/bases/s3.services.k8s.aws_buckets.yaml]: Yes this change was made by the latest version of `controller-gen`

File: test/e2e/cross-account/smoke.sh (hunk 7 above)
Comment (A-Hilaly) [reply test/e2e/cross-account/smoke.sh]: ++

File: pkg/runtime/reconciler.go [hunk 9]
```diff
@@ -106,14 +107,20 @@ func (r *reconciler) reconcile(req ctrlrt.Request) error {
 
 	acctID := r.getOwnerAccountID(res)
 	region := r.getRegion(res)
+	roleARN, _ := r.cache.Accounts.GetAccountRoleARN(string(acctID))
```
Comment (jaypipes) [inline pkg/runtime/reconciler.go:110]: How about we hide the cache implementation further and create a `reconciler.getRoleARN()` function similar to the `getOwnerAccountID` and `getRegion` functions.

File: pkg/runtime/reconciler.go [hunk 10]
```diff
@@ -106,14 +107,20 @@ func (r *reconciler) reconcile(req ctrlrt.Request) error {
...
+	sess, err := NewSession(region, ackv1alpha1.AWSResourceName(roleARN))
+	if err != nil {
+		return err
+	}
 
 	r.log.WithValues(
 		"account_id", acctID,
+		"assumed_role_arn", roleARN,
```
Comment (jaypipes) [inline pkg/runtime/reconciler.go:118]: just call it `role` instead of `assumed_role_arn`. And let's call the `account_id`log message key just `account`.

**Review by jaypipes — CHANGES_REQUESTED** [review]
two tiny requests inline. :)

File: pkg/runtime/reconciler.go (hunk 9 above)
Comment (A-Hilaly) [reply pkg/runtime/reconciler.go]: +1 That's a good idea! Done

File: pkg/runtime/reconciler.go (hunk 10 above)
Comment (A-Hilaly) [reply pkg/runtime/reconciler.go]: Done!

File: test/e2e/s3/cross-account.sh [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+source "$SCRIPTS_DIR/lib/common.sh"
+source "$SCRIPTS_DIR/lib/k8s.sh"
+source "$SCRIPTS_DIR/lib/testutil.sh"
+
+
+# AWS_ACCOUNT_ID_ALT should be configured to allow cross account resources management
+# from AWS_ACCOUNT_ID
+AWS_ACCOUNT_ID_ALT=${AWS_ACCOUNT_ID_ALT:-""}
+AWS_PROFILE_ALT=${AWS_PROFILE_ALT:-""}
```
Comment (jaypipes) [inline test/e2e/s3/cross-account.sh:15]: please work with @smuralee to get the self-hosted runners set up with an appropriate alternate profile/account ID for the e2e tests that will run on new PRs. :)

**Review by jaypipes — APPROVED** [review]
\o/

File: test/e2e/s3/cross-account.sh (hunk 11 above)
Comment (A-Hilaly) [reply test/e2e/s3/cross-account.sh:15]: Sure! Will reach out ASAP.
